src/System.Net.Security/src/System/Net/Security/SslState.cs

```diff
@@ -1279,18 +1271,38 @@ internal void FinishRead(byte[] renegotiateBuffer)
             }
         }
 
-        // Returns: 
-        // true  - operation queued
-        // false - operation can proceed
-        internal bool CheckEnqueueWrite(AsyncProtocolRequest asyncRequest)
+        internal Task CheckEnqueueWriteAsync()
+        {
+            //Clear previous request.
+            int lockState = Interlocked.CompareExchange(ref _lockWriteState, LockWrite, LockNone);
+            if (lockState != LockHandshake)
+            {
+                return Task.CompletedTask;
+            }
+            lock (this)
+            {
+                if (_lockWriteState != LockHandshake)
+                {
+                    CheckThrow(true);
+                    return Task.CompletedTask;
+                }
+
+                _lockWriteState = LockPendingWrite;
+                TaskCompletionSource<int> completionSource = new TaskCompletionSource<int>();
+                _queuedWriteStateRequest = completionSource;
```

Comment:
Rather than allocating a TaskCompletionSource for each of these, consider using `AsyncTaskMethodBuilder`.  It's a struct, so won't require an allocation each time.  You just need to make sure that you access its Task property before any other thread could try to SetResult/Exception on it, as Task lazily-initializes the task.  You also need to be careful about not copying it, as it's a mutable struct.

Reply:
Oh, I just realized _queuedWriteStateRequest is typed as object rather than as `TaskCompletionSource<int>`.  Nevermind then, since the `ATMB` would just get boxed anyway.